### PR TITLE
Callback Handler for MLflow

### PR DESCRIPTION
**Pull request by akshaya-a**
Rebased Mahmedk's PR with the callback refactor and added the example requested by hwchase plus a couple minor fixes

### CODE REVIEW

**Comment by akshaya-a** [comment]
@mahmedk if you don't mind, I didn't have write permissions on your fork so I added the sample notebook while preserving your commits so you get the credit in the commit history :)

@hwchase17 this includes the sample notebook you requested and I've tested it works with the local MLflow tracking store

File: langchain/chains/base.py [hunk 1]
```diff
@@ -268,7 +268,7 @@ def dict(self, **kwargs: Any) -> Dict:
         """Return dictionary representation of chain."""
         if self.memory is not None:
             raise ValueError("Saving of memory is not yet supported.")
-        _dict = super().dict()
+        _dict = super().dict(exclude={"callbacks"})
```
Comment (agola11) [inline langchain/chains/base.py:271]: this should be unnecessary after https://github.com/hwchase17/langchain/pull/4364

File: langchain/callbacks/__init__.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 from langchain.callbacks.stdout import StdOutCallbackHandler
 from langchain.callbacks.streaming_aiter import AsyncIteratorCallbackHandler
 from langchain.callbacks.wandb_callback import WandbCallbackHandler
+from langchain.callbacks.mlflow_callback import MlflowCallbackHandler
 
 __all__ = [
     "OpenAICallbackHandler",
```
[the remark this answers is not in the document]
Comment (akshaya-a) [reply langchain/callbacks/__init__.py:17]: indeed - bad merge thanks

File: langchain/chains/base.py (hunk 1 above)
Comment (akshaya-a) [reply langchain/chains/base.py:271]: great thanks! BTW for simplicity i kept the agent saving logic caught by the NotImplementedError but Agents + outputparser seem to have a similar serialization issue.

**Comment by akshaya-a** [comment]
@agola11  thanks for the review and apologies for not following the contributions instructions earlier - I've run the make lint + format now so things should be happy (and removed my change from Chain base).

After this PR, I am willing to add a few more tests in a follow up - are you willing to add mlflow + azureml-mlflow packages as test dependencies or should we explore some other options?

File: langchain/callbacks/mlflow_callback.py [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+from langchain.utils import get_from_dict_or_env
+
+
+def import_mlflow() -> Any:
+    try:
+        import mlflow
+    except ImportError:
+        raise ImportError(
+            "To use the mlflow callback manager you need to have the `mlflow` python "
+            "package installed. Please install it with `pip install mlflow==2.2.2`"
```
Comment (dbczumar) [inline langchain/callbacks/mlflow_callback.py:28]:
```suggestion
            "package installed. Please install it with `pip install mlflow>=2.3.0`"

```

**Review by dbczumar — APPROVED** [review]
LGTM with a nit comment: https://github.com/hwchase17/langchain/pull/4150/files#r1190588611. Thanks @akshaya-a @hwchase17 !

**Review by dbczumar — APPROVED** [review]
LGTM with a nit comment: https://github.com/hwchase17/langchain/pull/4150/files#r1190588611. Thanks @akshaya-a @hwchase17 !

**Comment by mahmedk** [comment]
> @mahmedk if you don't mind, I didn't have write permissions on your fork so I added the sample notebook while preserving your commits so you get the credit in the commit history :) EDIT: and I've added you as a collaborator on my fork in case you want to update it!
> 
> @hwchase17 this includes the sample notebook you requested and I've tested it works with the local MLflow tracking store

@akshaya-a thanks much ! Sorry, I could not attend to this work for past couple of weeks due to internal commitments. Glad to see the PR getting merged and mlflow support in langchain. Thanks Again :-)

**Pull request merged**